### PR TITLE
Ravil/gpu merge

### PR DESCRIPTION
**Pull request by ravil-mobile**
This request is the first stage of merging the GPU branch
* Updated reference to yateto
* Extended cmake input parameters
* Merged with Lukas's cmake branch
* Fixed Travis CI issues with cmake

### CODE REVIEW

File: .travis.yml
Comment (krenzland) [inline .travis.yml:71]: Is there a reason why you have to set +u at the end? I'd assume that it's not doing anything here?

File: Documentation/compilation.rst
Comment (krenzland) [inline Documentation/compilation.rst:10]: Should we maybe require gcc >= 8 here? As this one is used by travis.

File: Documentation/compilation.rst
Comment (krenzland) [inline Documentation/compilation.rst:15]: If cmake >3.15 is needed, this should be stated in the root level cmakefile -> otherwise it will fail with weird errors

File: Documentation/compilation.rst
Comment (krenzland) [inline Documentation/compilation.rst:42]: This uses cmake 3.16 - but documentation says 3.15 is needed. Maybe we should decide on one of both for docs/travis

File: cmake/FindGemmTools.cmake
Comment (krenzland) [inline cmake/FindGemmTools.cmake:53]: It would be better style to use https://cmake.org/cmake/help/v3.16/module/FindPython.html for this and specify that we want to have python3. Your version would use python2 on most systems. (I assume gemmforge is python3, as python2 is basically dead)

File: cmake/cpu_arch_flags.cmake
Comment (krenzland) [inline cmake/cpu_arch_flags.cmake:53]: Should be -march=power9 (as on all archs except on arm!). Or is this similar to arm? In this case -mcpu should be used...

File: cmake/process_users_input.cmake
Comment (krenzland) [inline cmake/process_users_input.cmake:32]: Spacing is really strange here

File: cmake/process_users_input.cmake
Comment (krenzland) [inline cmake/process_users_input.cmake:38]: Please avoid introducing new aligned lines, as changing them destroys the git diffs

File: cmake/process_users_input.cmake
Comment (krenzland) [inline cmake/process_users_input.cmake:81]: Maybe eigen should be the default here as it works on all platforms?

File: generated_code/aderdg.py
Comment (krenzland) [inline generated_code/aderdg.py:216]: Typo in method name

File: generated_code/aderdg.py
Comment (krenzland) [inline generated_code/aderdg.py:43]: We should try and avoid from common import *, especially if we are only importing one function. Better to stay explicit when importing something into the global space.

Comment (krenzland) [reply generated_code/aderdg.py:43]: Same of course for other instances of this line.

File: generated_code/aderdg.py
Comment (krenzland) [inline generated_code/aderdg.py:187]: Shouldn't platforms be a class member? In this case we do not need to pass it around every time. Mabye also true for the generator...

File: generated_code/aderdg.py
Comment (krenzland) [reply generated_code/aderdg.py:216]: Not sure if I like that pattern though -> it leads to a lot of duplicate code. Don't have a better idea right now...

File: generated_code/generate.py
Comment (krenzland) [inline generated_code/generate.py:94]: host_arch should be set outside of the if clause

File: src/Initializer/GlobalData.cpp
Comment (krenzland) [inline src/Initializer/GlobalData.cpp:52]: Where does float_t here actually come from?

Comment (krenzland) [reply src/Initializer/GlobalData.cpp:52]: And: What's the difference to real that we already have?

**Review by krenzland — CHANGES_REQUESTED** [review]
Looks good mostly! Note: I did not comment on my own cmake changes, someone else should probably do that

File: .travis.yml
Comment (ravil-mobile) [reply .travis.yml:71]: It is actually needed. *u* in the following line _set -euo pipefail_ forces bash to treat unset variables and parameters other than the special parameters ‘@’ or ‘*’ as an error when performing parameter expansion.There are some unused parameters in bash functions used by Travis. Therefore, I had to restore the settings back. This Travis bug has been reported and mentioned in some threads dedicated to Travis CI. Currently, it is the solution.

File: cmake/process_users_input.cmake
Comment (ravil-mobile) [reply cmake/process_users_input.cmake:32]: But it is easy to understand which alignment belongs to which architecture. It is done on purpose

File: cmake/process_users_input.cmake
Comment (ravil-mobile) [reply cmake/process_users_input.cmake:38]: I didn't get you

File: .travis.yml
Comment (krenzland) [reply .travis.yml:71]: I see. I knew what +euo pipefail does - I never write a bash script without, as not setting it can hide errors. Didn't know that that breaks travis. Thanks for figuring this out!

**Comment by Thomas-Ulrich** [comment]
Hi, Just trying to build this branch on Shaheen and getting:

```
CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
   Could NOT find ParMETIS (missing: PARMETIS_TEST_RUNS) (found version "4.0")
 Call Stack (most recent call first):
   /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
   cmake/FindParMETIS.cmake:120 (find_package_handle_standard_args)
   CMakeLists.txt:218 (find_package)
```
I run the following command:
```
ccmake .. -DCMAKE_PREFIX_PATH=/sw/xc40cle7/parmetis/4.0.3/cle7_gnu8.3.0;/sw/xc40cle7/metis/5.1.0/cle7_gnu8.3.0/
```
With the following module loaded:
```
x_ulricht@cdl2:/project/k1343/ulrich/SeisSol/build_cmake> module display metis
-------------------------------------------------------------------
/sw/xc40cle7/modulefiles/metis/5.1.0:

module-whatis    METIS : Serial Graph Partitioning and Fill-reducing Matrix Ordering 
module           swap PrgEnv-gnu PrgEnv-gnu 
setenv           METIS_DIR /sw/xc40cle7/metis/5.1.0/cle7_gnu8.3.0 
prepend-path     PATH /sw/xc40cle7/metis/5.1.0/cle7_gnu8.3.0/bin 
prepend-path     MANPATH /sw/xc40cle7/metis/5.1.0/cle7_gnu8.3.0/share/man 
prepend-path     LD_LIBRARY_PATH /sw/xc40cle7/metis/5.1.0/cle7_gnu8.3.0/lib 
append-path      INCLUDE /sw/xc40cle7/metis/5.1.0/cle7_gnu8.3.0/include 
-------------------------------------------------------------------
x_ulricht@cdl2:/project/k1343/ulrich/SeisSol/build_cmake> module display parmetis/4.0.3 
-------------------------------------------------------------------
/sw/xc40cle7/modulefiles/parmetis/4.0.3:

module-whatis    ParMETIS : Parallel Graph Partitioning and Fill-reducing Matrix Ordering 
setenv           PARMETIS_DIR /sw/xc40cle7/parmetis/4.0.3/cle7_gnu8.3.0 
prepend-path     PATH /sw/xc40cle7/parmetis/4.0.3/cle7_gnu8.3.0/bin 
prepend-path     MANPATH /sw/xc40cle7/parmetis/4.0.3/cle7_gnu8.3.0/share/man 
prepend-path     LD_LIBRARY_PATH /sw/xc40cle7/parmetis/4.0.3/cle7_gnu8.3.0/lib 
append-path      INCLUDE /sw/xc40cle7/parmetis/4.0.3/cle7_gnu8.3.0/include 
-------------------------------------------------------------------
````
I wonder if I m doing something wrong or not?

File: generated_code/common.py
Comment (krenzland) [inline generated_code/common.py:2]: typo in name -> l in kernel is missing

**Review by krenzland — CHANGES_REQUESTED** [review]
Would be great if you actually look at the proposed changes... most were from my last review.

**Comment by krenzland** [comment]
> Hi, Just trying to build this branch on Shaheen and getting:
> PS: The problem was that libmetis was not detected. I then needed to manually set it to: /sw/xc40cle7/metis/5.1.0/cle7_gnu8.3.0/lib/libmetis.so

Did this work on master? I don't see anything here that actually changes how metis is handled. I don't have access on Shaheen and can't reproduce.

**Comment by Thomas-Ulrich** [comment]
> > Hi, Just trying to build this branch on Shaheen and getting:
> > PS: The problem was that libmetis was not detected. I then needed to manually set it to: /sw/xc40cle7/metis/5.1.0/cle7_gnu8.3.0/lib/libmetis.so
> 
> Did this work on master? I don't see anything here that actually changes how metis is handled. I don't have access on Shaheen and can't reproduce.

Hi, I used the branch because yaml-cpp and ImpalaJIT were not built, and I ve seen that this branch compiles also these libs.
But no, this does not work either on master.

File: generated_code/aderdg.py
Comment (uphoffc) [reply generated_code/aderdg.py:219]: This is an abomination. We already have a concept to distinguish architectures which is the Arch class -- you just need to call generate twice giving a CPU Arch object in one call and a GPU Arch object in another call.

This also relates to the comment in the yateto PR: Why don't you make "target" part of Arch?

File: cmake/FindGemmTools.cmake
Comment (ravil-mobile) [reply cmake/FindGemmTools.cmake:53]: Ideally, we need to check a presence of python3 at top of the main CMake file.

File: generated_code/aderdg.py
Comment (ravil-mobile) [reply generated_code/aderdg.py:219]: Not all tensor expression must be generated for GPU inside a kernel. Consider Dynamic rupture: only 1 out of 3 kernels can be generated for GPU.

**Comment by krenzland** [comment]
>  But no, this does not work either on master.

Okay, then I'd propose we postpone this issue. I'll add it to my todo-list and come back to you once this PR is merged/NG is up again.

File: Documentation/compilation.rst
Comment (krenzland) [inline Documentation/compilation.rst:42]: The whole command should be set in () brackets. This opens a subshell - the user remains in the same directory.

**Review by krenzland — CHANGES_REQUESTED** [review]
Looks mostly good.
I still really dislike the loop over targets inside of each generator. Imo it should be outside and each generator should be called with one target.

**Review by krenzland — APPROVED** [review]

**Pull request merged**
